Multiple results for a resultspec, filename ran-on sorting for efficiency, 'hdat' entrypoint

dillonwilliams opened the pull request:
Also, the "report" command was removed per previous discussion. I haven't checked getting multiple results on the command-line, but it seems like it could use a dialog like "Continue to show/diff next result for <suite>/<case> [Y/n]?" 

Closes #11, closes #3.

dillonwilliams committed:
Remove "report" command

The "show" command could be used to tabulate and report results when used for a suite.

dillonwilliams committed:
Implement resultspecs for multiple results

Also tests for relative result indexing. Multiple result handling needs to be tested interactively; we may want dialogs for this. Still depickling for timestamps.

dillonwilliams committed:
Use filename/result_id to sort results by run time

Previously results had to be depickled to get ran_on time (still used as fallback). Overriding result_id generation in user suites is also no longer supported.

dillonwilliams committed:
Add 'hdat' console entry point in setup.py

main.py is now the global entry point. The hdat command can be tested with "pip install -e .  && hdat", still possible to run with ./hdat/main.py

dillonwilliams committed:
Add case directory to args tuple for suite recents

For Python 3.4, where only named arguments can follow *expression

dillonwilliams committed:
Use listdir instead of scandir for python 3.4

os.scandir is 2-3x faster for checking properties but not in 3.4

johndgiese commented on hdat/hdat_cli.py:
```diff
@@ -0,0 +1,113 @@
+
+    diff_help = 'compare two results'
+    diff_parser = subparsers.add_parser('diff', help=diff_help)
+    diff_golden_help = 'result to compare to (defaults to current golden for the case)'
```
I just noticed this, but in order to do a diff, we need the context, but the context for the current golden set likely won't be available.

I think the easiest solution is to remove the default behavior and make the golden resultspec mandatory.

dillonwilliams replied:
Kind of confused about this one.

johndgiese replied:
Sorry, let me expand on my point a bit more.

The CLI currently accepts two arguments.  The first is `goldenspec`, which is a resultspec that should resolve to X results.  The second is `resultspec`, which is also a resultspec that should resolve to Y results.  We expect that X == Y, and should print an error if they do not.  Then, we do something like:

```
for x, y in zip(X, Y):
  suite.diff(x, y)
```

dillonwilliams replied:
New syntax in latest commit is simply `hdat diff <resultspec> <resultspec>`

johndgiese replied:
Nice, looks like `git diff` :)

![screen shot 2018-02-15 at 2 22 45 pm](https://user-images.githubusercontent.com/1648754/36276533-4cdf2d94-125c-11e8-95c3-944b3e0b2d6e.png)

johndgiese commented on hdat/hdat_cli.py:
```diff
@@ -0,0 +1,113 @@
+
+    if args.command == 'run' and args.collect_only:
+        cases = resolve_casespecs(suites, args.casespecs)
+        print("\n".join(['{}/{}'.format(s, c) for s, c in cases]))
```
I realize this was leftover from my code, but when I was reading this, I couldn't remember what `s` and `c` were, or the format that `resolve_casespecs` returns.  I think my code would have been better if I used:

```
print("\n".join(['{}/{}'.format(suite_id, case_id) for suite_id, case_id in cases]))
```

what do you think?

dillonwilliams replied:
Agreed

johndgiese commented on hdat/hdat_cli.py:
```diff
@@ -0,0 +1,113 @@
+    elif args.command == 'show':
+        results = resolve_resultspecs(archive, args.resultspec)
+        for result in results:
+            show_result(suites, result)
```
It seems like we don't need a `y/n` prompt built into hdat, because we rely on users to stop the for loop in their `show` methods.  The problem with putting the prompt into `hdat` is that it forces it on users.

We actually don't want a prompt because matplotlib blocks anyway.

Also, I can definitely imagine cases where you would want to display all of the figures for all the matching result specs at once, without having to step through them one by one.

johndgiese commented on hdat/hdat_cli.py:
```diff
@@ -0,0 +1,113 @@
+        results = resolve_resultspecs(archive, args.resultspec)
+        for result in results:
+            show_result(suites, result)
+    elif args.command == 'runshow':
```
When I added the `runshow` command, that I couldn't think of as simple way to avoid duplication between `run`, `show`, and `runshow`.  That said, I suspect there is a good way to do it.

dillonwilliams replied:
The duplication isn't so bad, but the commands could be broken out from the if/elif blocks into their own functions.

johndgiese commented on hdat/hdat_cli.py:
```diff
@@ -0,0 +1,113 @@
+        cases_status = run_cases(suites, golden_store, archive, git_info, cases)
+        if cases_status['error'] > 0:
+            raise AbortError(_format_cases_status(cases_status))
+        results = resolve_resultspecs(archive, args.casespecs[0])
```
This is probably a leftover bug of mine, but I think `args.casespecs[0]` should be `args.casespecs`.

dillonwilliams replied:
👍

johndgiese commented on hdat/main.py:
```diff
@@ -1,132 +1,41 @@
-import argparse
-import traceback
-from collections import OrderedDict
+#!/usr/bin/env python3
```
👍

johndgiese commented on hdat/store.py:
```diff
@@ -17,21 +20,54 @@ def select(self, suite_id, case_id, result_id):
-        for filename in result_filenames:
-            result = self.read_result(filename)
-            results.append(result)
+        ResultDesc = namedtuple('ResultDesc', ('id', 'ran_on'))
```
Perhaps we should define this outside of the class?  I guess it is nice having it closer to the code that uses it, but its less efficient (but efficiency probably doesn't matter much here)...

dillonwilliams replied:
This is mostly for readability vs. a normal tuple here. Agree if this sorting case will show up elsewhere.

johndgiese commented on setup.py:
```diff
@@ -47,5 +47,9 @@
     package_data={},
     data_files=[],
-    entry_points={},
+    entry_points={
```
👍

johndgiese commented on tests/resultspec_test.py:
```diff
@@ -44,7 +44,7 @@ def archive(tmp_archive, mock_results):
     def test_existing_file(self, archive, mock_results):
         resultspec = archive._result_filename('a', '1', 'r1')
-        assert resolve_resultspec(archive, resultspec) == mock_results[0]
+        assert resolve_resultspecs(archive, resultspec)[0] == mock_results[0]
```
Would `assert resolve_resultspecs(archive, resultspec) == mock_results` work?

dillonwilliams replied:
There should only be one result coming back from resolve_resultspecs in this case, but this isn't obvious from the test. I can clear this up.

johndgiese replied:
Ahh, I see.

In that case, I think:

```
assert resolve_resultspecs(archive, resultspec) == mock_results[:1]
```

Would be better than

```
assert resolve_resultspecs(archive, resultspec)[0] == mock_results[0]
```

since it would ensure that the length of the return value is correct, in addition to the first items matching.

johndgiese commented on hdat/store.py:
```diff
@@ -17,21 +20,54 @@ def select(self, suite_id, case_id, result_id):
-        return results_sorted
+        results = []
+        for entry in os.listdir(top_directory):
+            if not entry.startswith('.') and os.path.isdir(os.path.join(top_directory, entry)):
```
So `os.listdir` skips over `.` and `..`, so is the `startswith` guard for other hidden directories?

https://docs.python.org/3/library/os.html#os.listdir

dillonwilliams replied:
Correct, and the os.scandir documentation (which is not in 3.4) says that a lot of the performance increase comes from not having to check dir vs. file again after getting the directory entries.

johndgiese commented on hdat/store.py:
```diff
@@ -17,21 +20,54 @@ def select(self, suite_id, case_id, result_id):
+        return results
+
+    def select_recents_all(self):
+        results = []
```
If the context files are large (or if there are a lot of them), then we may use up quite a bit of RAM here.

Maybe we should return a generator instead?

dillonwilliams replied:
Definitely, I'll try this out.

johndgiese reviewed (commented):
Looks great! Nice work; I left a few comments

dillonwilliams committed:
Clarity improvements for tests and hdat_cli

dillonwilliams committed:
Use generators for returning multiple results

dillonwilliams committed:
Use two resultspecs on 'hdat diff', removing 'goldenspec'

Also fixed bug in comparing suite_id and added test for 'hdat diff'

dillonwilliams committed:
Rename suite.verify to suite.check

Checks are automatic during runs; 'hdat verify' manually updates a golden result
